### DISCUSSION TITLE
Library for painless and powerful API route middleware

### DISCUSSION THREAD

**Discussion by htunnicliff** [vercel/next.js#19721] [Show and tell]
**Hey Next.js community 👋!**

Using middleware in Next.js API routes has always vexed me — why couldn't the simple middleware patterns from Express (or even better, Koa) be brought into Next applications? 

After writing far too many workarounds in my own applications to get middleware working well, I decided to put some effort into making a library ([`next-api-middleware`](https://github.com/htunnicliff/next-api-middleware)) to solve the problem! 

<p>
  <a href="https://github.com/htunnicliff/next-api-middleware">
    <img src="https://forthebadge.com/images/badges/check-it-out.svg">
  </a>
  <a href="https://github.com/htunnicliff/next-api-middleware">
    <img src="https://forthebadge.com/images/badges/compatibility-betamax.svg">
  </a>
  <a href="https://github.com/htunnicliff/next-api-middleware">
    <img src="https://forthebadge.com/images/badges/uses-badges.svg">
  </a>
</p>

Here's how it works:

```ts
import { use } from "next-api-middleware";

const middleware = use(
  async (req, res, next) => {
    console.log("Do work before the request");
    await next();
    console.log("Clean up");
  },

  async (req, res, next) => {
    console.log("Do more work");
    await next();
    console.log("Clean up more");
  },

  (req, res, next) => {
    console.log("Store user in request");
    req.locals = {
      user: {
        name: "Alice",
        email: "alice@example.com",
      },
    };

    return next();
  }
);

const apiRouteHandler = async (req, res) => {
  const { name } = req.locals.user;

  res.status(200);
  res.send(`Hello, ${name}!`);
};

export default middleware(apiRouteHandler);

/**
 * Console output:
 *
 *   Do work before the request
 *   Do more work
 *   Store user in request
 *   Clean up more
 *   Clean up
 */
```

The library exposes two primary methods – `use` and `label` – for easily composing groups of middleware. Please give this library and try and let me know what you think!

**Comment by junhsss** [comment]
It elegantly solved my problem. You made my day. Thanks! ☺️
